R/ops.R

```diff
@@ -142,65 +69,268 @@ Ops.tf <- function(e1, e2) {
+         `*`   = ,
+         `/`   = ,
+         `^`   = ,
+         `%%`  = ,
+         `%/%` = tfb_op_tfb(op, x, y),
+         stop_incompatible_op(op, x, y)
+  )
+}
+
+#' @export
+#' @method vec_arith.tfb numeric
+vec_arith.tfb.numeric <- function(op, x, y, ...) {
+  switch(op,
+         `/`   = ,
+         `*`   = tfb_multdiv_numeric(op, x, y),
+         `+`   = ,
+         `-`   = ,
+         `^`   = ,
+         `%%`  = ,
+         `%/%` = tfb_op_numeric(op, x, y),
+         stop_incompatible_op(op, x, y)
+  )
+}
+
+#' @export
+#' @method vec_arith.numeric tfb
+vec_arith.numeric.tfb <- function(op, x, y, ...) {
+  switch(op,
+         `*` = tfb_multdiv_numeric(op, y, x),
+         `+`   = ,
+         `-`   = ,
+         `/`   = ,
+         `^`   = ,
+         `%%`  = ,
+         `%/%` = numeric_op_tfb(op, x, y),
+         stop_incompatible_op(op, x, y)
+  )
+}
+
+#' @export
+#' @method vec_arith.tfb MISSING
+vec_arith.tfb.MISSING <- function(op, x, y, ...) {
+  tf_op_missing(op, x, y, ...)
+}
+
+#-------------------------------------------------------------------------------
+
+tfd_op_tfd <- function(op, x, y) {
+  assert_compatible_size(op, x, y)
+  # TODO: could be more lenient -- allow is one domain is subset of the other?
+  same_domain <- all.equal(tf_domain(x), tf_domain(y), check.attributes = FALSE) |>
+    isTRUE()
+  same_arg <- all.equal(tf_arg(x), tf_arg(y), check.attributes = FALSE) |>
+    isTRUE()
+  if (!same_domain || !same_arg) {
+    message <- c(
+      glue::glue("<{vec_ptype_full(x)}> {op} <{vec_ptype_full(y)}>  not permitted for different",
+                 "{ifelse(same_domain, '', ' domains')}",
+                 "{ifelse(!same_domain & !same_arg, ' and', '')}",
+                 "{ifelse(same_arg, '', ' argument values')}"),
+        "-- use tf_rebase first.") |>
+      paste(collapse = "\n")
+    stop_incompatible_op(op, x, y, message = message)
+  }
+
+  attr_ret <- if (vec_size(x) >= vec_size(y)) attributes(x) else attributes(y)
+  arg_ret <- tf_arg(y)
+  x_ <- tf_evaluations(x)
+  y_ <- tf_evaluations(y)
+  ret <- map2(x_, y_, \(x, y) do.call(op, list(x, y)))
+
+  if (attr(x, "evaluator_name") != attr(y, "evaluator_name")) {
+    warning(
+      "inputs have different evaluators, result has ", attr_ret$evaluator_name,
+      call. = FALSE
+    )
+  }
+  if ("tfd_irreg" %in% attr_ret$class) {
+    ret <- map2(arg_ret, ret, \(x, y) list(arg = x, value = y))
+  }
+  attributes(ret) <- attr_ret
+  ret
+}
+
+tfd_op_numeric <- function(op, x, y, ...) {
+  assert_compatible_size(op, x, y)
+  ret <- map2(tf_evaluations(x), y, \(x, y) do.call(op, list(x, y)))
+  if (is_irreg(x)) {
+    ret <- map2(tf_arg(x), ret, \(.arg, .ret) list(arg = .arg, value = .ret))
+  }
+  attributes(ret) <- attributes(x)
+  if (vec_size(y) > 1) {
+    names(ret) <- NULL
+  }
+  ret
+}
+
+# some code-duplication here, this makes non-commutative ops work for tfd and numeric
+numeric_op_tfd <- function(op, x, y) {
```

Comment:
Is this code-duplication required, i.e. might the implementation differ in the future?